File: control-plane/src/main/java/it/unimib/datai/nanofaas/controlplane/config/CoreDefaults.java
package it.unimib.datai.nanofaas.controlplane.config;

import it.unimib.datai.nanofaas.controlplane.registry.ImageValidator;
import it.unimib.datai.nanofaas.controlplane.service.InvocationEnqueuer;
import it.unimib.datai.nanofaas.controlplane.service.ScalingMetricsSource;
import it.unimib.datai.nanofaas.controlplane.sync.SyncQueueGateway;
import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.context.annotation.Fallback;

@Configuration
public class CoreDefaults {

    @Bean
    @ConditionalOnMissingBean(InvocationEnqueuer.class)
    public InvocationEnqueuer invocationEnqueuer() {
        return InvocationEnqueuer.noOp();
    }

    @Bean
    @ConditionalOnMissingBean(ScalingMetricsSource.class)
    public ScalingMetricsSource scalingMetricsSource() {
        return ScalingMetricsSource.noOp();
    }

    @Bean
    @ConditionalOnMissingBean(SyncQueueGateway.class)
    public SyncQueueGateway syncQueueGateway() {
        return SyncQueueGateway.noOp();
    }

    @Bean
    @Fallback
    @ConditionalOnMissingBean(ImageValidator.class)
    public ImageValidator imageValidator() {
        return ImageValidator.noOp();
    }
}

File: experiments/control-plane-staging/versions/control-plane-rust-m3-20260222-200159/snapshot/control-plane-rust/src/core_defaults.rs
use crate::registry::{ImageValidator, NoOpImageValidator};
use crate::service::{InvocationEnqueuer, NoOpInvocationEnqueuer, NoOpScalingMetricsSource, ScalingMetricsSource};
use crate::sync::{NoOpSyncQueueGateway, SyncQueueGateway};
use std::sync::Arc;

pub struct CoreDefaults {
    pub invocation_enqueuer: Arc<dyn InvocationEnqueuer + Send + Sync>,
    pub scaling_metrics_source: Arc<dyn ScalingMetricsSource + Send + Sync>,
    pub sync_queue_gateway: Arc<dyn SyncQueueGateway + Send + Sync>,
    pub image_validator: Arc<dyn ImageValidator + Send + Sync>,
}

impl CoreDefaults {
    pub fn new(
        invocation_enqueuer: Option<Arc<dyn InvocationEnqueuer + Send + Sync>>,
        scaling_metrics_source: Option<Arc<dyn ScalingMetricsSource + Send + Sync>>,
        sync_queue_gateway: Option<Arc<dyn SyncQueueGateway + Send + Sync>>,
        image_validator: Option<Arc<dyn ImageValidator + Send + Sync>>,
    ) -> Self {
        Self {
            invocation_enqueuer: invocation_enqueuer
                .unwrap_or_else(|| Arc::new(NoOpInvocationEnqueuer)),
            scaling_metrics_source: scaling_metrics_source
                .unwrap_or_else(|| Arc::new(NoOpScalingMetricsSource)),
            sync_queue_gateway: sync_queue_gateway
                .unwrap_or_else(|| Arc::new(NoOpSyncQueueGateway)),
            image_validator: image_validator.unwrap_or_else(|| Arc::new(NoOpImageValidator)),
        }
    }
}
